src/cpu/SearchBlock.zig

```diff
@@ -0,0 +1,344 @@
+        }
+    }
+    return out;
+}
+
+const Coord = struct { x: usize, z: usize };
+
+/// add preprocessed value to count
+const use_preprocessed = blk: {
+    var buf: [30]Coord = undefined;
+    var coords = std.ArrayListUnmanaged(Coord).initBuffer(&buf);
+    for (mask, 0..) |row, x| {
+        for (row, 0..) |char, z| {
+            if (char == '+' or char == '-') coords.appendAssumeCapacity(.{ .x = x, .z = z });
+        }
+    }
+    const coords_final = coords.items[0..coords.items.len];
+    break :blk coords_final.*;
+};
+
+/// add slime value of cell to count
+const add = blk: {
+    var buf: [30]Coord = undefined;
+    var coords = std.ArrayListUnmanaged(Coord).initBuffer(&buf);
+    for (mask, 0..) |row, x| {
+        for (row, 0..) |char, z| {
+            if (char == 'o') coords.appendAssumeCapacity(.{ .x = x, .z = z });
+        }
+    }
+    const coords_final = coords.items[0..coords.items.len];
+    break :blk coords_final.*;
+};
+
+/// subtract binary value of cell from count
+const sub = blk: {
+    var buf: [30]Coord = undefined;
+    var coords = std.ArrayListUnmanaged(Coord).initBuffer(&buf);
+    for (mask, 0..) |row, x| {
+        for (row, 0..) |char, z| {
+            if (char == '-') coords.appendAssumeCapacity(.{ .x = x, .z = z });
+        }
+    }
+    const coords_final = coords.items[0..coords.items.len];
+    break :blk coords_final.*;
+};
+
+/// For every chunk within the searched area defined by this `SearchBlock`
+/// checks whether the amount of slime chunk slime chunks in spawn range of a player
+/// at the center of each chunk meets the given `threshold`
+pub fn calculateSliminess(
+    self: *@This(),
+    params: slimy.SearchParams,
+    context: anytype,
+    comptime resultCallback: fn (@TypeOf(context), slimy.Result) void,
+) usize {
+    var sufficiently_slimy_chunks: usize = 0;
+    for (0..size - mask.len + 1) |x| {
+        for (0..size - mask.len + 1) |z| {
+            var count: u8 = 0;
+
+            var preprocessed_count_1: u16 = 0;
+            inline for (use_preprocessed[0..15]) |location| {
+                preprocessed_count_1 += @as(u8, @bitCast(self.data[x * size + z + location.x * size + location.z]));
+            }
+            var preprocessed_count_2: u16 = 0;
+            inline for (use_preprocessed[15..]) |location| {
+                preprocessed_count_2 += @as(u8, @bitCast(self.data[x * size + z + location.x * size + location.z]));
+            }
+            count += @intCast((preprocessed_count_1 >> 4) + (preprocessed_count_2 >> 4));
+
+            var add_count: u16 = 0;
+            inline for (add) |location| add_count += @as(u8, @bitCast(self.data[x * size + z + location.x * size + location.z]));
+            count += @intCast(add_count & 0xf);
+
+            var sub_count: u16 = 0;
+            inline for (sub) |location| sub_count += @as(u8, @bitCast(self.data[x * size + z + location.x * size + location.z]));
+            count -= @intCast(sub_count & 0xf);
```

Comment:
i'm not 100% following this code - what's the reason for using bitwise ops instead of field access?

Also `inline for` is rarely appropriate for optimization purposes - have you benchmarked this?

Reply:
The `.strip_count` field is stored in the upper 4 bits of the Cell struct. The other field is `.slime`, which is `u4` for padding purposes but is actually stores `bool`.
To access `.strip_count`, zig/llvm shifts the backing int of Cell (u8) left: >> 4. However, since `.slime` is guaranteed to be 0 or 1, we can just directly add the backing ints together up to 15 times before overflow occurs and shift >> 4 at the end. Since `use_preprocessed` contains more than 15 elements, we split it into two sub-arrays. This optimization saves 24 (26 - 2) right shifts, and combined with similar optimizations for `add_count` and `sub_count`, speed up the function by over 80%.
I tried signalling to LLVM that the padding bits are always 0 with asserts, but it was unable to perform the optimization. I discussed it with qwerasd in the Zig discord and they said, after experimentation, "it seems like it does not have any concept of 'the lower bits will never overflow in to this area, so we don't need to pre-shift'".

Reply:
As for the `inline for` - I never benchmarked it😅, and after doing so it doesn't seem to make any difference. I'll remove it now